invoices/syncing/tenkfeet.py

```diff
@@ -210,119 +238,189 @@ def match_user(self, email):
+                "category": entry.get("task") or "[none]",
+                "notes": entry.get("notes"),
+                "entry_type": "Confirmed" if not entry["is_suggestion"] else "Suggestion",
+                "discipline": user.get("discipline") or "[none]",
+                "role": user.get("role") or "[none]",
+                "bill_rate": entry["bill_rate"],
+                "leave_type": legacy_leave_type(),
+                "phase_name": assignable.get("phase", {}).get("phase_name") or "[Non Phase Specific]",
+                "billable": is_phase_billable(
+                    phase_name="",
+                    project=project.get("name") or "[Leave Type]"
+                ),  # TODO
+                "approved": approval is not None,
+                "status": status(approval),
+                "user_email": user["email"].lower(),
+                "last_updated_at": now,
+                "calculated_is_billable": is_phase_billable(
+                    phase_name=assignable.get("phase", {}).get("phase_name") or "[Non Phase Specific]",
+                    project=project.get("name") or "[Leave Type]"
+                ),
+                "upstream_approvable_id": approval.get("id"),
+                "upstream_approvable_updated_at": approval.get("updated_at"),
+            }
+
+            assert data["date"].year > 2000
+            assert data["date"].year < 2050
+            assert data["bill_rate"] >= 0
+            assert data["incurred_money"] >= 0
+            assert data["incurred_hours"] >= 0
+
+            # Reset billing rates and money for all leaves
+            if data["leave_type"] != "[project]":
+                data["incurred_money"] = data["bill_rate"] = 0
+
+            return data
+
         self.logger.info("Starting hour entry update: %s - %s", self.start_date, self.end_date)
-        now = timezone.now()
-        tenkfeet_entries = tenkfeet_api.fetch_hour_entries(self.start_date, self.end_date)
-        tenkfeet_api_entries = {entry["id"]: entry for entry in tenkfeet_api.fetch_api_hour_entries(self.start_date, self.end_date)}
-        entries = []
 
-        per_date_data = defaultdict(lambda: {"items": [], "sha256": hashlib.sha256()})
+        users = {user.user_id: model_to_dict(user) for user in TenkfUser.objects.all()}
 
-        for entry in tenkfeet_entries:
-            if parse_float(entry[8]) in (0, None):  # incurred_hours
-                logger.debug("Skipping hour entry with 0 incurred hours: %s", entry)
-                continue
-            entry_date = parse_date(entry[40])
-            entry_data = {"reporting": entry, "api": tenkfeet_api_entries.get(parse_upstream_id(entry[59]), {})}
-            per_date_data[entry_date]["sha256"].update(json.dumps(entry_data).encode())
-            per_date_data[entry_date]["items"].append(entry_data)
+        self.logger.info("Fetch assignables.")
+        assignables = fetch_assignables()  # TODO cache this
```

Comment:
Does it make sense to catch this or `fetch_phases()`?

Reply:
Would be nice, yes. Required, not really. Could easily load that data when synching projects.